File: decorator/c++/milk-decorator.cpp
#include "milk-decorator.h"

MilkDecorator::MilkDecorator(Coffee* decoratedCoffee) : CoffeeDecorator(decoratedCoffee) {};

double MilkDecorator::getCost() {
    return CoffeeDecorator::getCost() + 1;
}

std::string MilkDecorator::getIngredients() {
    return CoffeeDecorator::getIngredients() + " Milk";
}

File: decorator/js/milk-decorator.js
exports.decorator = function(coffee) {
    var originalCost = coffee.getCost;
    var originalIngredients = coffee.getIngredients;

    coffee.getCost = function() {
        return originalCost.apply(this, arguments) + 1;
    }
    coffee.getIngredients = function() {
        return originalIngredients.apply(this, arguments) + " Milk";
    }

    return coffee;
}